include/deal.II/lac/vector.templates.h

```diff
@@ -614,6 +563,498 @@ namespace internal
+
+    RealType
+    operator() (size_type &i) const
+    {
+      return numbers::NumberTraits<Number>::abs(X[i++]);
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(size_type &i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      i += VectorizedArray<Number>::n_array_elements;
+      return std::abs(x);
+    }
+
+    const Number *X;
+  };
+
+  template <typename Number, typename RealType>
+  struct NormP
+  {
+    static const bool vectorizes = types_are_equal<Number,double>::value ||
+                                   types_are_equal<Number,float>::value;
+
+    RealType
+    operator() (size_type &i) const
+    {
+      return std::pow(numbers::NumberTraits<Number>::abs(X[i++]), p);
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(size_type &i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      i += VectorizedArray<Number>::n_array_elements;
+      return std::pow(std::abs(x),p);
+    }
+
+    const Number *X;
+    RealType p;
+  };
+
+  template <typename Number>
+  struct MeanValue
+  {
+    static const bool vectorizes = types_are_equal<Number,double>::value ||
+                                   types_are_equal<Number,float>::value;
+
+    Number
+    operator() (size_type &i) const
+    {
+      return X[i++];
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(size_type &i) const
+    {
+      VectorizedArray<Number> x;
+      x.load(X+i);
+      i += VectorizedArray<Number>::n_array_elements;
+      return x;
+    }
+
+    const Number *X;
+  };
+
+  template <typename Number>
+  struct AddAndDot
+  {
+    static const bool vectorizes = types_are_equal<Number,double>::value ||
+                                   types_are_equal<Number,float>::value;
+
+    Number
+    operator() (size_type &i) const
+    {
+      X[i] += a * V[i];
+      Number xi = X[i];
+      return xi * Number(numbers::NumberTraits<Number>::conjugate(W[i++]));
+    }
+
+    VectorizedArray<Number>
+    do_vectorized(size_type &i) const
+    {
+      VectorizedArray<Number> x, w, v;
+      x.load(X+i);
+      v.load(V+i);
+      x += a * v;
+      x.store(X+i);
+      // may only load from W after storing in X because the pointers might
+      // point to the same memory
+      w.load(W+i);
+      i += VectorizedArray<Number>::n_array_elements;
+      return x * w;
+    }
+
+    Number *X;
+    const Number *V, *W;
+    Number a;
+  };
+
+
+
+  // this is the inner working routine for the accumulation loops
+  // below. This is the standard case where the loop bounds are known. We
+  // pulled this function out of the regular accumulate routine because we
+  // might do this thing vectorized (see specialized function below)
+  template <typename Operation, typename ResultType>
+  void
+  accumulate_regular(const Operation &op,
+                     size_type       &n_chunks,
+                     size_type       &index,
+                     ResultType (&outer_results)[128],
```

Comment:
128 is hard-coded (we could make it a constant throughout this file if we want). This pre-dates this commit and is selected because we have hand-coded loops up to size 4096 (to avoid to go down too deeply with the recursive calls). The last 32 (=4x8) items are unrolled in loops as you see in the body, so you are left with 128 items that you combine and sum over later. So essentially this is necessary to get the pairwise summation without recursion.